### PR TITLE
feat(APIv2): RHINENG-2875 implement tailorings endpoint

### PR DESCRIPTION
**Pull request by romanblanco**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: app/models/v2/tailoring.rb [hunk 1]
```diff
@@ -7,15 +7,35 @@ class Tailoring < ApplicationRecord
...
     validates :profile, presence: true
     validates :os_minor_version, numericality: { greater_than_or_equal_to: 0 }
+
+    def os_major_version
+      attributes['security_guide__os_major_version'] || security_guide.os_major_version
+    end
+
+    def value_overrides
+      attributes['profile__value_overrides'] || profile.value_overrides
+    end
```
Comment (skateman) [inline app/models/v2/tailoring.rb:39]: No, `value_overrides` is part of the tailoring not the profile in this scope. Each tailoring has its own set of rules and values, only the default values are coming from the parent.

File: app/policies/v2/policy_policy.rb [hunk 2]
```diff
@@ -19,7 +19,7 @@ def destroy?
       match_account?
     end
 
-    # Only show hosts in our user account
+    # Only show policy in our user account
```
Comment (skateman) [inline app/policies/v2/policy_policy.rb:22]: Out of scope of this PR, please extract it

File: app/policies/v2/tailoring_policy.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    def index?
+      true
+    end
+
+    def show?
+      true
+    end
+
+    def tailoring_file?
+      true # TODO: cert_auth
```
Comment (skateman) [inline app/policies/v2/tailoring_policy.rb:15]: Please create a ticket for it if you are not planning to solve it in this PR.

File: app/policies/v2/tailoring_policy.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+  # Only show tailoring in our user account
+  class Scope < V2::ApplicationPolicy::Scope
+    def resolve
+      resolve scope.where('1=0') if user&.account_id.blank?
+
+      scope
+        .where
+        .associated(:policy)
+        .where(policy: { account_id: user.account_id })
```
Comment (skateman) [inline app/policies/v2/tailoring_policy.rb:27]: Does this have test coverage?

File: app/serializers/v2/tailoring_serializer.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+module V2
+  # JSON API serialization for an OpenSCAP Profile tailoring
+  class TailoringSerializer < V2::ApplicationSerializer
+    attributes :profile_id
+    attributes :policy_id
+    attributes :os_minor_version
+
+    derived_attribute :os_major_version, security_guide: [:os_major_version]
+    derived_attribute :value_overrides, profile: [:value_overrides]
```
Comment (skateman) [inline app/serializers/v2/tailoring_serializer.rb:11]: As I stated above, this should not be a derived attribute.

File: app/services/v2/xccdf_tailoring_file.rb [hunk 6]
```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+module V2
+  # A class representing an XCCDF Tailoring File
+  class XccdfTailoringFile
```
Comment (skateman) [inline app/services/v2/xccdf_tailoring_file.rb:5]: As most of this is copy-paste, I'd recommend to revisit this for refactoring after the tests are ready.

File: config/environments/development.rb [hunk 7]
```diff
@@ -62,6 +62,7 @@
   else # Impose a 64MB limit for the logfile
     config.logger = Insights::Api::Common::LoggerWithAudit.new(config.paths['log'].first, 1, 64.megabytes)
   end
+  ActiveRecord::Base.logger = Logger.new STDOUT
```
Comment (skateman) [inline config/environments/development.rb:65]: 🔥 🚒 ⚡

File: config/routes.rb [hunk 8]
```diff
@@ -34,7 +34,11 @@ def draw_routes(prefix)
             end
           end
 
-          resources :policies, except: [:new, :edit]
+          resources :policies, except: [:new, :edit] do
+            resources :tailorings, only: [:index, :show], parents: [:policy] do
+              get 'tailoring_file', on: :member
```
Comment (skateman) [inline config/routes.rb:39]: @RoamingNoMaD I think you can use the `on: :collection` similarly in your PR 🤔 not having a strong opinion about it, but agree with @romanblanco on some kind of convention.

File: spec/api/v2/schemas.rb [hunk 9]
```diff
@@ -6,13 +6,17 @@ module Api
   module V2
     module Schemas
       include Metadata
+      include Policies
```
Comment (skateman) [inline spec/api/v2/schemas.rb:9]: Please extract this into a separate PR, I understand it is required, but still out of scope.

File: spec/controllers/v2/tailorings_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+    it_behaves_like 'collection', :policy
+    it_behaves_like 'paginable', :policy
+    it_behaves_like 'searchable', :policy
+    it_behaves_like 'sortable', :policy
+  end
+
+  describe 'GET show' do
+    let(:profile) { FactoryBot.create(:v2_profile) }
+    let(:parent) do
```
Comment (skateman) [inline spec/controllers/v2/tailorings_controller_spec.rb:52]: Please do logical separation of lazy variables, at least the long ones.

File: spec/controllers/v2/tailorings_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+      FactoryBot.create(
+        :v2_policy,
+        account: current_user.account,
+        profile: profile
+      )
+    end
+    let(:item) do
```
Comment (skateman) [inline spec/controllers/v2/tailorings_controller_spec.rb:59]: This is very disturbing to read 😕

File: spec/controllers/v2/tailorings_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    let(:extra_params) { { policy_id: parent.id, id: item.id } }
+    let(:notfound_params) do
+      extra_params.merge(policy_id: FactoryBot.create(
+        :v2_policy,
+        account: current_user.account,
+        profile: FactoryBot.create(:v2_profile)
+      ).id)
+    end
+
+    it_behaves_like 'individual', :policy
```
Comment (skateman) [inline spec/controllers/v2/tailorings_controller_spec.rb:75]: You only have test for individual behavior, but collections should be tested as well. Definitely for sorting/searching...

File: spec/factories/rule.rb [hunk 13]
```diff
@@ -16,6 +16,16 @@
       profile_id { nil }
     end
 
+    trait :with_group do
+      rule_group { association :v2_rule_group, security_guide: security_guide }
+    end
+
+    trait :with_group_hierarchy do
+      rule_group do
+        association :v2_rule_group, parent_count: (1..5).to_a.sample, security_guide: security_guide
```
Comment (skateman) [inline spec/factories/rule.rb:25]: Can you please extract `parent_count` into a transient variable? We should avoid using constants in factories.

File: spec/factories/rule_group.rb [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    end
+
+    ancestry do
+      if parent_count == 1
+        create(:v2_rule_group).id
+      elsif parent_count > 1
+        parent_group = create(:v2_rule_group, parent_count: parent_count - 1)
+        [parent_group.ancestry, parent_group.id].join('/')
+      end
+    end
```
Comment (skateman) [inline spec/factories/rule_group.rb:22]: I guess this would need some explanation in comments...

File: spec/factories/tailoring.rb [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+          ref_id: ref_id,
+          profile_os_minor_versions: {
+            os_minor_version: os_minor_version
+          }
+        )
+    end
+
+    transient do
+      ref_id { policy.profile.ref_id }
+      value_overrides { {} }
```
Comment (skateman) [inline spec/factories/tailoring.rb:18]: I guess this should be copied from the `profile.value_overrides` as a default value.

File: spec/factories/tailoring.rb [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+    after(:create) do |tailoring, _|
+      tailoring.profile.rules.each do |rule|
+        FactoryBot.create(
+          :v2_tailoring_rule,
+          rule: rule,
+          tailoring: tailoring
+        )
+      end
+    end
```
Comment (skateman) [inline spec/factories/tailoring.rb:30]: I think we would need a trait about randomly shuffled rules where not only the rules of the parent profile are mixed in, but other ones from the security guide as well.

File: spec/factories/tailoring_rule.rb [hunk 17]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :v2_tailoring_rule, class: 'V2::TailoringRule' do
+  end
```
Comment (skateman) [inline spec/factories/tailoring_rule.rb:5]:
```suggestion
  factory :v2_tailoring_rule, class: 'V2::TailoringRule'
```

File: spec/fixtures/files/searchable/tailorings_controller.yaml [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+  :entities:
+    :found:
+      - :factory: :v2_tailoring
+        :os_minor_version: 8
+        :policy_id: ${policy_id}
+    :not_found:
+      - :factory: :v2_tailoring
+        :os_minor_version: 1
+        :policy_id: ${policy_id}
+  :query: (os_minor_version != 1)
```
Comment (skateman) [inline spec/fixtures/files/searchable/tailorings_controller.yaml:24]: This file is never being used as there is no `searchable` in the controller spec.

File: spec/fixtures/files/sortable/tailorings_controller.yaml [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+    :policy_id: ${policy_id}
+
+:queries:
+  - :sort_by:
+      - 'os_minor_version:asc'
+    :result: [0, 2, 1]
+
+  - :sort_by:
+      - 'os_minor_version:desc'
+    :result: [1, 2, 0]
```
Comment (skateman) [inline spec/fixtures/files/sortable/tailorings_controller.yaml:21]: This file is not used as there is no `sortable` in the spec.

**Review by skateman — CHANGES_REQUESTED** [review]

File: spec/api/v2/schemas.rb (hunk 9 above)
Comment (RoamingNoMaD) [reply spec/api/v2/schemas.rb:9]: @romanblanco If you agree, I'd copy this into my [rswag PR](https://github.com/RedHatInsights/compliance-backend/pull/1868) for now. There it wouldn't be out of scope.

Comment (romanblanco) [reply spec/api/v2/schemas.rb:9]: @RoamingNoMaD Yes, that would be better :+1:

File: spec/controllers/v2/tailorings_controller_spec.rb (hunk 11 above)
Comment (romanblanco) [reply spec/controllers/v2/tailorings_controller_spec.rb:59]: Not sure what you mean here, are you asking for a space in between?

Comment (skateman) [reply spec/controllers/v2/tailorings_controller_spec.rb:59]: Yes

File: spec/controllers/v2/tailorings_controller_spec.rb [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+
+  describe 'GET tailoring_file' do
+    let(:parent) do
+      canonical_profiles = [6, 7, 8].map do |version|
+        FactoryBot.create(
+          :v2_profile,
+          :with_rules,
+          os_major_version: 8,
+          ref_id_suffix: 'foo',
+          supports_minors: [version]
```
Comment (romanblanco) [inline spec/controllers/v2/tailorings_controller_spec.rb:86]: note: We can support multiple minor RHEL versions.
